### PR TITLE
Improve nix setup + fix minor install script typo

### PR DESCRIPTION
**Pull request by imacg**
* Rewrite `default.nix`
* Modify dev scripts to invoke `bash` through `/usr/bin/env` (required for systems that don't put `bash` in `/bin/bash`)
* add a `.envrc` that invokes nix to integrate with editors
* Fix minor issue in `quick-install/install.sh` where its banner unintentionally invokes `gcloud`

### CODE REVIEW

**Comment by amouat** [comment]
This looks great thanks!

Regarding the banner, I guess I never noticed, as I have gcloud installed.

I'll take a closer look later but I don't see any issues atm.

**Comment by blaggacao** [comment]
Just a friendly pointer to an extremely worthwhile project in this space: https://github.com/numtide/devshell

File: shell.nix [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    (rustChannelOf { inherit date; channel = "nightly"; }).rust
+
+    pkgconfig
+    cmake
+    perl
+    go
+    sqlite
+    protobuf
+    openssl
+    valgrind
```
Comment (blaggacao) [inline shell.nix:27]:
```suggestion
    valgrind
    kustomize
```

Comment (blaggacao) [inline shell.nix:27]:
```suggestion
    valgrind
    kustomize
    k9s
    kubeprompt
    dive
    stern
    kube3d
```
A selection of what we use and what might be useful...
<details>
<summary>Our full devshell entrypoint for reference</summary>

```nix
{ src
, preCommitExcludes ? []
, repoSpecificShellHook ? ""
}:
let
  nixpkgs-src = import ./nixpkgs-src.nix;
  overlay = import ./overlays/overlay.nix;
  pkgs = import nixpkgs-src {
    config = {};
    overlays = [ overlay ];
  };

  pre-commit = (
    import ./pre-commit.nix {
      inherit pkgs src;
      excludes = preCommitExcludes;
    }
  );
  devshell = pkgs.callPackage ( import ./devshell-src.nix ) { };

  # Overlays
  customTools = with pkgs; [
    j2cli
    commitizen
    container-structure-test
    kube-score
    skaffold-nix-builder
  ];

  nixTooling = with pkgs; [
    morph # gitops for nix (better UX, stateless)
    nix-diff # explains why two Nix derivations differ
    nix # nix package manager -- tagged version
    nixops # gitops for nix (stateful)
    nixos-container # run nixos in a systemd nspawn container (don't confuse with OCI compliant containers)
    nixpkgs-fmt # format nix files canonically
    nix-prefetch # determine the hash of a fixed-output derivation (TOFU)
    nix-top # help users figure out what's building
    nix-tree # interactively browse a Nix store paths dependencies
  ];

  gitAndRepoTooling = with pkgs; [
    bat # beat cat
    cfssl # generate certificates
    drone-cli # run CI locally
    fd # faster than find
    fzf # fuzzy find
    git
    gitAndTools.delta # better cli diffs
    gitAndTools.git-filter-repo # rewrite git history (eg extract subfolders)
    git-town # branching workflows
    hostctl # manage /etc/hosts file with style
    just # task runner
    kubeprompt # isolate kubconfigs in a shell
    nodePackages.prettier # auto format lots of things
    ripgrep # rest in peace grep!
    scc # calculate project cost / time COCOMO model
    shfmt # format shell files
    yq-go # yaml / json cli manipulator, akin to jq
  ];

  containerTooling = with pkgs; [
    dive
    k9s
    kpt
    kube3d
    kubectl
    kubernetes-helm
    kubeval
    kustomize
    lazydocker
    podman
    skaffold
    stern
  ];

in
devshell {
  name = "${baseNameOf src}";
  packages = nixTooling ++ gitAndRepoTooling ++ containerTooling ++ customTools;
  env = {
    # Fix the nixpkgs version to devshell's version
    # See: https://github.com/numtide/devshell/issues/26
    # NIX_PATH = "nixpkgs=${pkgs.path}";

    # common kube settings
    KUBECTL_EXTERNAL_DIFF = "delta";

    # common k9s settings
    K9SCONFIG = "${toString ./k9s.d}";

		# Skaffold environment for K8sApp based dev environments (not relevant in neptun)
		# TODO: secure MTL & CFX regitries
		SKAFFOLD_INSECURE_REGISTRY = "registry.quantumops.rocks:5000";
  };
  bash = {

    extra = ''
    playground_zone_kubekonfig="$HOME/.kube/git.p1.rocks/gitops/DevShell/playground-zone"
    k3d_playground_kubeconfig="$HOME/.kube/git.p1.rocks/gitops/DevShell/k3d-playground"
    export KUBECONFIG="$KUBECONFIG:$HOME/.kube/config:$playground_zone_kubekonfig:$k3d_playground_kubeconfig";
    # git town settings
    git config git-town.code-hosting-driver gitea
    git config git-town.code-hosting-origin-hostname git.p1.rocks
    git config git-town.main-branch-name master

    # git delta settings
    git config core.pager delta
    git config interactive.diffFilter "delta --color-only"
    git config diff.colorMoved default
    git config delta.navigate true
    git config delta.side-by-side true
    git config delta.line-numbers true
    git config delta.decorations true
    git config delta.whitespace-error-style "22 reverse"
    ${repoSpecificShellHook}
    ${pre-commit.shellHook}

    # zsh prompt configuration
    export IN_WHICH_NIX_SHELL="$(basename $(pwd))"
    export KUBEPROMPT_FORMAT='{{if .Enabled}}%B%F{blue}%{⎈ %2G%}%f➜%{ %}%F{yellow}{{.Ctx}}%f|%F{magenta}{{.Ns}}%f%b{{end}}'
    export RPROMPT='$(kubeprompt -f default)'

    # Fix the nixpkgs version to devshell's version
    export NIX_PATH="nixpkgs=${pkgs.path}"
    '';

    interactive = ''

    '';
  };
}

```

</details>

**Comment by imacg** [comment]
@blaggacao thanks for pointing out devshell and your example config, I'll definitely check them both out in depth.

With respect to the current PR it's my understanding that Nix hasn't been used as a development tool within this project for awhile. My aim is to keep things simple and workable for my own purposes while minimizing the overhead on the maintainers (that don't use Nix).

**Comment by amouat** [comment]
Yes, that's accurate @imacg. I don't use nix and can't really test this, but I'm very happy to include it if it doesn't affect others.

@imacg let me know when you're happy with the PR and I'll merge.

**Comment by blaggacao** [comment]
Just wanto to let you know that I'm planning on working on firat class support for _hostctl_ and _mkcert_ to nix's improved devshell today because I need to adopt trow in a "link-local" development setup through those tools.

Not meaning to hold back merging, though. An eventual later refactoring is straight forward.

**Comment by imacg** [comment]
@amouat I'm good with the current PR.

**Review by amouat — APPROVED** [review]

**Comment by amouat** [comment]
@imacg This PR qualifies for our Hacktoberfest campaign: https://blog.container-solutions.com/container-solutions-does-hacktoberfest

If you'd like a t-shirt, could you e-mail me at adrian.mouat@container-solutions.com with your size and address?

**Comment by amouat** [comment]
Hi @imacg - I don't think you sent me an e-mail? If you'd like a t-shirt please send me an e-mail with your size and address.

Either way, thanks for the PR :)

**Pull request merged**